bfdd/bfdd_northbound.c

```diff
@@ -0,0 +1,1118 @@
+/*
+ * BFD daemon northbound implementation.
+ *
+ * Copyright (C) 2019 Network Device Education Foundation, Inc. ("NetDEF")
+ *                    Rafael Zalamena
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301 USA.
+ */
+
+#include "lib/log.h"
+#include "lib/northbound.h"
+
+#include "bfd.h"
+
+/*
+ * Prototypes.
+ */
+void bfd_session_get_key(bool mhop, const struct lyd_node *dnode,
+			 struct bfd_key *bk);
+int bfd_session_create(enum nb_event event, const struct lyd_node *dnode,
+		       union nb_resource *resource, bool mhop);
+int bfd_session_destroy(enum nb_event event, const struct lyd_node *dnode,
+			bool mhop);
+
+/*
+ * Helpers.
+ */
+void bfd_session_get_key(bool mhop, const struct lyd_node *dnode,
+			 struct bfd_key *bk)
+{
+	const char *ifname = NULL, *vrfname = NULL;
+	struct sockaddr_any psa, lsa;
+
+	/* Required destination parameter. */
+	strtosa(yang_dnode_get_string(dnode, "./dest-addr"), &psa);
+
+	/* Get optional source address. */
+	memset(&lsa, 0, sizeof(lsa));
+	if (yang_dnode_exists(dnode, "./source-addr"))
+		strtosa(yang_dnode_get_string(dnode, "./source-addr"), &lsa);
+
+	/* Get optional interface and vrf names. */
+	if (yang_dnode_exists(dnode, "./interface"))
+		ifname = yang_dnode_get_string(dnode, "./interface");
+	if (yang_dnode_exists(dnode, "./vrf"))
+		vrfname = yang_dnode_get_string(dnode, "./vrf");
+
+	/* Generate the corresponding key. */
+	gen_bfd_key(bk, &psa, &lsa, mhop, ifname, vrfname);
+}
+
+int bfd_session_create(enum nb_event event, const struct lyd_node *dnode,
+		       union nb_resource *resource, bool mhop)
+{
+	struct bfd_session *bs;
+	struct bfd_key bk;
+
+	switch (event) {
+	case NB_EV_VALIDATE:
+		bfd_session_get_key(mhop, dnode, &bk);
+		bs = bfd_key_lookup(bk);
+
+		/* This session was already configured by CLI. */
+		if (bs != NULL && BFD_CHECK_FLAG(bs->flags, BFD_SESS_FLAG_CONFIG))
+			return NB_ERR_VALIDATION;
```

Comment:
Is this really necessary? YANG shouldn't allow us to create duplicate list entries with identical keys.